src/firmware_new/boards/frankie_v1/io/drivetrain.c

```diff
@@ -1,48 +1,56 @@
 #include "firmware_new/boards/frankie_v1/io/drivetrain.h"
 
-#include <assert.h>
-#include <stdbool.h>
-
-static DrivetrainUnit_t *_front_left_drive_unit;
-static DrivetrainUnit_t *_front_right_drive_unit;
-static DrivetrainUnit_t *_back_left_drive_unit;
-static DrivetrainUnit_t *_back_right_drive_unit;
-
-static bool initialized = false;
-
-void io_drivetrain_init(DrivetrainUnit_t *front_left_drive_unit,
-                        DrivetrainUnit_t *front_right_drive_unit,
-                        DrivetrainUnit_t *back_left_drive_unit,
-                        DrivetrainUnit_t *back_right_drive_unit)
+struct Drivetrain
 {
-    _front_left_drive_unit  = front_left_drive_unit;
-    _front_right_drive_unit = front_right_drive_unit;
-    _back_left_drive_unit   = back_left_drive_unit;
-    _back_right_drive_unit  = back_right_drive_unit;
-
-    initialized = true;
-}
-
-void io_drivetrain_applyForceFrontLeftWheel(float force_newtons)
+    DrivetrainUnit_t* front_left_drive_unit;
+    DrivetrainUnit_t* front_right_drive_unit;
+    DrivetrainUnit_t* back_left_drive_unit;
+    DrivetrainUnit_t* back_right_drive_unit;
+    Matrix robot_wheel_geometry;
+};
+
+DrivetrainUnit_t* io_drivetrain_init(DrivetrainUnit_t* front_left_drive_unit,
+                                     DrivetrainUnit_t* front_right_drive_unit,
+                                     DrivetrainUnit_t* back_left_drive_unit,
+                                     DrivetrainUnit_t* back_right_drive_unit,
+                                     Matrix robot_wheel_geometry)
 {
-    assert(initialized);
-    io_drivetrain_unit_applyForce(_front_left_drive_unit, force_newtons);
+    Drivetrain_t* drive_train = (Drivetrain_t*)malloc(sizeof(Drivetrain_t));
+
+    drive_train->front_left_drive_unit  = front_left_drive_unit;
+    drive_train->front_right_drive_unit = front_right_drive_unit;
+    drive_train->back_left_drive_unit   = back_left_drive_unit;
+    drive_train->back_right_drive_unit  = back_right_drive_unit;
+    drive_train->robot_wheel_geometry   = robot_wheel_geometry;
 }
 
-void io_drivetrain_applyForceFrontRightWheel(float force_newtons)
+void io_drivetrain_destroy(Drivetrain_t* drivetrain)
 {
-    assert(initialized);
-    io_drivetrain_unit_applyForce(_front_right_drive_unit, force_newtons);
+    free(drivetrain);
 }
 
-void io_drivetrain_applyForceBackLeftWheel(float force_newtons)
+static Matrix io_drivetrain_localRobotSpeed2WheelSpeeds(Drivetrain_t* drive_train,
```

Comment:
```suggestion
static Matrix io_drivetrain_localRobotSpeedToWheelSpeeds(Drivetrain_t* drive_train,
```
would prefer this naming